Fixes #5079 - :authority header for IPv6 address not having square br…

sbordet opened the pull request:
…ackets.

On the client:
* Origin.Address.host is passed through HostPort.normalizeHost(),
so that if it is IPv6 is bracketed.
Now the ipv6 address passed to an `HttClient` request is bracketed.
* HttpRequest was de-bracketing the host, but now it does not anymore.

On the server:
* Request.getLocalAddr(), getLocalName(), getRemoteAddr(),
getRemoteHost(), getServerName(), when dealing with an IPv6 address,
return it bracketed.
The reason to return bracketed IPv6 also from *Addr() methods is that
if it is used with InetAddress/InetSocketAddress it still works, but
often it is interpreted as a URI host so brackets are necessary.
* DoSFilter was blindly bracketing - now it does not.

Added a number of test cases, and fixed those that expected
non-bracketed IPv6.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5079 - :authority header for IPv6 address not having square brackets.

On the client:
* Origin.Address.host is passed through HostPort.normalizeHost(),
so that if it is IPv6 is bracketed.
Now the ipv6 address passed to an `HttClient` request is bracketed.
* HttpRequest was de-bracketing the host, but now it does not anymore.

On the server:
* Request.getLocalAddr(), getLocalName(), getRemoteAddr(),
getRemoteHost(), getServerName(), when dealing with an IPv6 address,
return it bracketed.
The reason to return bracketed IPv6 also from *Addr() methods is that
if it is used with InetAddress/InetSocketAddress it still works, but
often it is interpreted as a URI host so brackets are necessary.
* DoSFilter was blindly bracketing - now it does not.

Added a number of test cases, and fixed those that expected
non-bracketed IPv6.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime commented:
Some test failures in ...

* `org.eclipse.jetty.proxy.ForwardProxyServerTest`
* `org.eclipse.jetty.proxy.ForwardProxyTLSServerTest`

sbordet committed:
Merged branch 'jetty-9.4.x' into 'jetty-9.4.x-5079-ipv6_brackets'.

sbordet committed:
Fixes #5079 - :authority header for IPv6 address not having square brackets.

Fixed Jenkins failures by disabling tests that require IPv6 if it is not available.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java:
```diff
@@ -1155,10 +1155,14 @@ protected HttpField getAcceptEncodingField()
+     * @return the host itself
+     * @deprecated no replacement, do not use it
+     */
+    @Deprecated
```
should the impl at least call HostPort.normalize so that it works if somebody does call it?

joakime replied:
Also, perhaps add some apidoc text telling developers what the options are now that it's deprecated ...

``` java
/**
 * @deprecated use {@link HostPort#normalize(String)} instead.
 */
@Deprecated
```

sbordet replied:
Nope, the method should not be used. It's `protected` so only subclasses could use it, but they should not - there is no need to normalize the host anymore.

I initially followed @gregw suggestion to call `HostPort.normalizeHost()`, but actually this method was doing exactly the opposite. So either restore the old behavior, or keep this new one that does nothing.

gregw reviewed (approved):
other than a niggle LGTM

sbordet committed:
Fixes #5079 - :authority header for IPv6 address not having square brackets.

Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5079 - :authority header for IPv6 address not having square brackets.

Reverted code changes to HttpClient.normalizeHost().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
